\documentclass{article}
\usepackage[dvips]{graphicx,epsfig,color}
\title{User's manual of Self-organizing clustering}
\author{Kou AMANO \\ RIKEN BioResource Center}
\begin{document}
\maketitle

\section{Outline}
Self-organizing clustering (SOC) is a vector clustering machine which is based on k-means.
It provides 
1. function of conversion from DNA sequences to vectors, 
2. function of initialization of cluster configuration, 
3. function of cluster alignment with improved learning process.
Above the three functions are implemented as three commands ({\tt fasta2matrix}, {\tt soc-init} and {\tt soc-lm}).

\section{Install}
First, you have to obtain tar ball of SOC.
To obtain the source code, please contact {\tt <kamano@affrc.go.jp>} or {\tt <amano@brc.riken.jp>} .
To install SOC, please execute following steps:
\begin{verbatim}
[user]$ tar -zxvf SOC_0.5.3-00.tar.gz
[user]$ cd SOC_0.5.3-00
[user]$ ./configure.pl
[user]$ make
[user]$ su
[root]# mkdir -p /home/pub/bin
[root]# make install
[root]# exit
[user]$ PATH=$PATH:/home/pub/bin
[user]$ export PATH
\end{verbatim}

\section{Commands}
The latest version of SOC is 0.5.8-05.
\subsection{\tt fasta2matrix}
\subsubsection{Description}
{\tt fasta2matrix} command execute conversion from a set of multiple DNA sequence data to a float matrix data.
{\tt fasta2matrix} needs administrative information such as memory size.
Users have to specify memory sizes: {\tt <number of samples>} $\times$ ({\tt <header size> $+$ <body size>}) .
\subsubsection{Usage}
{\tt fasta2matrix [if=<file name>] [s=<segment size>] [g=<step size>]

[n=<number of samples>] [hs=<header (comment) size>]

[bs=<body (sequence) size>] [--help|-h] [--check|-c] [-H|+H]

[-m|+m] [-p<total count of signals>|+p] }

\

Options:
\begin{itemize}
\item {\tt <file name>} \\
To specify input file name.
\item {\tt <segment size>} \\
To specify length of oligonucleotides.
\item {\tt <step size>} \\
To specify window slide size.
\item {\tt <number of samples>} \\
For memory allocation.
{\tt fasta2matrix} stores all of multiple fasta sequence data on memory.
\item {\tt <header size>} \\
For memory allocation.
\item {\tt <body size>} \\
For memory allocation.
\item {\tt [--check|-c]} \\
To print command arguments.
\item {\tt [-H|+H]} \\
Where '-H' is specified, the program prints headers to output file.
\item {\tt [-m|+m]} \\
Where '-m' is specified, the program prints matrix size to output file.
\item {\tt <total count of signals>} \\
If users need oligonucleotide "ratio", the users can specify '-p' option.
\end{itemize}
\subsubsection{Examples}
To print oligonucleotide matrix with 100$\times$64 of size from file {\tt test.fasta}: \\
{\tt fasta2matrix if=test.fasta s=3 n=100}

\

To print oligonucleotide matrix with 100$\times$64 of size without entry name from file {\tt test.fasta}: \\
{\tt fasta2matrix if=test.fasta s=3 n=100 +H}

\

To print oligonucleotide matrix with 100$\times$64 of size without matrix size  from file {\tt test.fasta}: \\
{\tt fasta2matrix if=test.fasta s=3 n=100 +m}

\subsection{\tt soc-init}
\subsubsection{Description}
{\tt soc-init} provide several primitive types of initial cluster configuration.
\subsubsection{Usage}
{\tt soc-init if=<sample data file> of=<output data file>

[frac=<number of cluster nodes>] [<FORMAT>]}

\

Options:
\begin{itemize}
\item {\tt <sample data file>} \\
To specify the file of sample-nodes data for input.
\item {\tt <output data file>} \\
To specify the file of cluster-nodes data to output.
\item {\tt <number of cluster nodes>} \\
To specify number of initial cluster-nodes.
\item {\tt <FORMAT>} \\
See below.
\end{itemize}
{\tt <FORMAT>}:
\begin{itemize}
\item {\tt Diagonal} \\
To array cluster-nodes on a line across the sample vector space.
%\item {\tt random=Order}
%\item {\tt random=Value}
\item {\tt node=Central} \\
To place cluster-nodes at the sample-nodes near to the centroid.
%\item {\tt node=outLying}
%\item {\tt node=Intensive}
%\item {\tt node=Power}
%\item {\tt node=diStant}
%\item {\tt node=Median}
%\item {\tt node=Median,<n> (Under construction)}
\item {\tt Grid=<tensor size>} \\
To array cluster-nodes at network-nodes of grid.
\item {\tt Axis-mean=<axis1,axis2,...>} \\
To place cluster-nodes on coordinate axes of sample space.
%\item {\tt aXis-intensive=<ratio>}
\end{itemize}

\begin{tabular}{l}
\includegraphics[width=50mm]{soc-init-option.example.eps}
\begin{tabular}{cp{50mm}}
	\vspace{-49mm} & \\
	& Where {\tt Diagonal} option is specified, the cluster-nodes array as red dots.\\
	& Where {\tt node=Central}  option is specified, the cluster-nodes are placed on sample-nodes nearest to centroid (x). \\
	& Where {\tt Grid} option is specified, the cluster-nodes are placed on all of the lattice points.\\
	& Where {\tt Axis-mean} option is specified, the cluster-nodes array as blue dots.\\
	%\includegraphics[width=70mm]{init-options.eps} &   hoge hoge hoge hoge hoge hoge hoge hoge hoge hogehoge hoge hoge hoge hoge \\
	%\includegraphics[width=70mm, bb=0 100 200 200]{init-options.eps}
\end{tabular} \\
\multicolumn{1}{c}{Fig: Initialization patterns of cluster-nodes}
\end{tabular}

\subsubsection{Examples}
To locate 10 cluster-nodes around the centroid and write the coordinates to the file "test.fasta.mat.cl": \\
{\tt soc-init if=test.fasta.mat of=test.fasta.mat.cl frac=10 node=Central}

\

To locate 16 (4$\times$4) cluster-nodes on the nodes of the grid and write the coordinates to the file "test.fasta.mat.cl": \\
{\tt soc-init if=test.fasta.mat of=test.fasta.mat.cl Grid=4,4}

\subsection{\tt soc-lm}
\subsubsection{Description}
{\tt soc-lm}  learns to  move cluster-nodes, which represent clusters, expressing the centroids of clusters.
It dynamically generates/merges cluster-nodes in the learning stage.
\subsubsection{Usage}
{\tt soc-lm sf=<sample file> cf=<cluster file> [loop=<loop>]

[rate=<learning rate>] [unify=<n1>,<n2>,<n3>,<n4>]

[generate=<n1>,<n2>,<n3>,<n4>] [cutoff=<n1>,<n2>,<n3>]

[cutoff-gen=<n1>,<n2>,<n3>,<n4>] [delete=<n1>,<n2>,<n3>]

[lf=<log file>] [of=<output file>] [of\_type=<output file type>]

[rf=<report file>] [dist\_func=<distance function>]}

\

Options:
\begin{itemize}
\item {\tt <sample file>} \\
To specify sample file (input data).
\item {\tt <cluster file>} \\
To specify cluster file (input data) which is generated by {\tt soc-init}.
\item {\tt <loop>} \\
To specify learning iterations.
\item {\tt <learning rate>} \\
To specify learning rate to align cluster-nodes.
\item {\tt unify=<n1>,<n2>,<n3>,<n4>} \\
To specify the thresholds for unifying cluster-nodes. \\
IF {\tt n1} is specified AND distance between given cluster-node pair is lower than {\tt n2} AND number of clusters is higher than {\tt n3}, THEN the cluster-node pair whose distance is shortest is unified with interval {\tt n4}.
\item {\tt generate=<n1>,<n2>,<n3>,<n4>} \\
To specify the thresholds for generating cluster-node. \\
IF {\tt n1} is specified AND cluster-radius is higher than {\tt n2} AND number of clusters is lower than {\tt n3}, THEN the cluster-nodes whose radius is longest is divided with interval {\tt n4}.
%\item {\tt cutoff=<n1>,<n2>,<n3>} \\
%To specify the thresholds to decide outside member(s). \\
%IF {\tt n1} is specified AND distances between centroids to given cluster members are higher than {\tt n2} AND current loop iterations is higher than {\tt n3}, THEN the corresponding members are set on the "outside".
%{\tt n1}がセットされており、かつ、{\tt n2}を上回るクラスタノード-メンバ間の距離がある場合、{\tt n3}以降のループでは、当該のメンバの座標に新しいクラスタノードを発生させる。
%\item {\tt cutoff-gen=<n1>,<n2>,<n3>,<n4>} \\
%To specify the thresholds to decide outside of members and to generate new cluster(s). \\
%IF {\tt n1} is specified AND distances between centroids to given cluster members are higher than {\tt n2} AND distances are higher than {\tt n3}, THEN the corresponding members are set on the "outside" and generate cluster-nodes for the "outsiders" with interval {\tt n4}.
%\\[{\tt n1}がセットされており、かつ、{\tt n2}を上回るクラスタノード-メンバー間の距離があり、かつ、そのメンバーに"outside"のフラグが設定されている場合、そのメンバーをクラスタノードとする。]
\item {\tt delete=<n1>,<n2>,<n3>} \\
To specify the threshold to delete cluster-node(s). \\
IF {\tt n1} is specified AND distances between cluster-node pairs is higher than {\tt n2}, THEN ones of the pairs are deleted until all distances of cluster-node pair is shorter than {\tt n2} with interval {\tt n3}.
\\{\tt n1}がセットされており、かつ、{\tt n2}を上回るクラスタノード間の距離がある場合、{\tt n2}を上回る距離をもつクラスタノードペアがなくなるまで、当該のクラスタノードペアの片方をdeleteする。
\item {\tt <log file>} \\
To specify log file.
\item {\tt <output file>} \\
To specify output file.
\item {\tt <output file type>} \\
To specify output file type.
\item {\tt <report file>} \\
To specify another (reduced) output file.
\item {\tt <distance function>} \\
To specify distance function (cos$\theta$ or euclidean).
\end{itemize}
\subsubsection{Examples}
To classify sample data ({\tt test.fasta.mat}) with default of option: \\
{\tt soc-lm sf=test.fasta.mat cf=t.fasta.mat.cl}

\

To classify sample data ({\tt test.fasta.mat}) under the condition - loop is 100, cluster deletion is on, cluster deletion threshold is 4 and learning rate is 0.9: \\
{\tt soc-lm sf=test.fasta.mat cf=test.fasta.mat.cl delete=1,4,2 rate=0.9}
\end{document}
